src/cart/crt_context.c

```diff
@@ -1910,3 +1957,152 @@ crt_req_force_completion(struct crt_rpc_priv *rpc_priv)
 	crt_req_timeout_track(rpc_priv);
 	D_MUTEX_UNLOCK(&crt_ctx->cc_mutex);
 }
+
+static int
+context_quotas_init(crt_context_t crt_ctx)
+{
+	struct crt_context	*ctx = crt_ctx;
+	struct crt_quotas	*quotas;
+	int			rc = 0;
+
+	if (ctx == NULL) {
+		D_ERROR("NULL context\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	quotas = &ctx->cc_quotas;
+
+	quotas->limit[CRT_QUOTA_RPCS]   = crt_gdata.cg_rpc_quota;
+	quotas->current[CRT_QUOTA_RPCS] = 0;
+	quotas->enabled[CRT_QUOTA_RPCS] = crt_gdata.cg_rpc_quota > 0 ? true : false;
+out:
+	return rc;
+}
+
+static int
+context_quotas_finalize(crt_context_t crt_ctx)
+{
+	struct crt_context	*ctx = crt_ctx;
+
+	if (ctx == NULL) {
+		D_ERROR("NULL context\n");
+		return -DER_INVAL;
+	}
+
+	for (int i = 0; i < CRT_QUOTA_COUNT; i++)
+		ctx->cc_quotas.enabled[i] = false;
+
+	return DER_SUCCESS;
+}
+
+int
+crt_context_quota_limit_set(crt_context_t crt_ctx, crt_quota_type_t quota, int value)
+{
+	struct crt_context	*ctx = crt_ctx;
+	int			rc = 0;
+
+	if (ctx == NULL) {
+		D_ERROR("NULL context\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (quota < 0 || quota >= CRT_QUOTA_COUNT) {
+		D_ERROR("Invalid quota %d passed\n", quota);
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	D_MUTEX_LOCK(&ctx->cc_quotas.mutex);
+	ctx->cc_quotas.limit[quota] = value;
+	D_MUTEX_UNLOCK(&ctx->cc_quotas.mutex);
+
+out:
+	return rc;
+}
+
+int
+crt_context_quota_limit_get(crt_context_t crt_ctx, crt_quota_type_t quota, int *value)
+{
+	struct crt_context	*ctx = crt_ctx;
+	int			rc = 0;
+
+	if (ctx == NULL) {
+		D_ERROR("NULL context\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (quota < 0 || quota >= CRT_QUOTA_COUNT) {
+		D_ERROR("Invalid quota %d passed\n", quota);
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (value == NULL) {
+		D_ERROR("NULL value\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	*value = ctx->cc_quotas.limit[quota];
+
+out:
+	return rc;
+}
+
+int
+crt_context_get_quota_resource(crt_context_t crt_ctx, crt_quota_type_t quota)
+{
+	struct crt_context	*ctx = crt_ctx;
+	int			rc = 0;
+
+	if (ctx == NULL) {
+		D_ERROR("NULL context\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (quota < 0 || quota >= CRT_QUOTA_COUNT) {
+		D_ERROR("Invalid quota %d passed\n", quota);
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	/* If quotas not enabled or unlimited quota */
+	if (!ctx->cc_quotas.enabled[quota] || ctx->cc_quotas.limit[quota] == 0)
+		return 0;
+
+	D_MUTEX_LOCK(&ctx->cc_quotas.mutex);
+	if (ctx->cc_quotas.current[quota] < ctx->cc_quotas.limit[quota])
+		ctx->cc_quotas.current[quota]++;
+	else {
+		D_WARN("Quota limit reached for quota_type=%d\n", quota);
```

Comment:
This is overly chatty and should be a debug

Reply:
NB: "overly chatty" == gigabytes of client logs under load. ;)

Reply:
done